operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1791 @@
+	ciliumSvcLBIPSAnnotation = "io.cilium/lb-ipam-ips"
+
+	// The string used in the FieldManager field on update options
+	ciliumFieldManager = "cilium-operator-lb-ipam"
+
+	serviceNamespaceLabel = "io.kubernetes.service.namespace"
+	serviceNameLabel      = "io.kubernetes.service.name"
+)
+
+var Cell = cell.Module(
+	"lbipam",
+	"LB-IPAM",
+	// Provide LBIPAM so instances of it can be used while testing
+	cell.Provide(newLBIPAM),
+	// Invoke an empty function which takes an LBIPAM to force its construction.
+	cell.Invoke(func(*LBIPAM) {}),
+)
+
+type LBIPAMParams struct {
+	cell.In
+
+	Logger logrus.FieldLogger
+
+	LC         hive.Lifecycle
+	Shutdowner hive.Shutdowner
+
+	Clientset    k8sClient.Clientset
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	SvcResource  resource.Resource[*slim_core_v1.Service]
+
+	DaemonConfig *option.DaemonConfig
+}
+
+func newLBIPAM(params LBIPAMParams) *LBIPAM {
+	var lbClasses []string
+	if params.DaemonConfig.EnableBGPControlPlane {
+		lbClasses = append(lbClasses, "io.cilium/bgp-control-plane")
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+
+	lbIPAM := &LBIPAM{
+		ctx:          ipamCtx,
+		cancel:       cancel,
+		logger:       params.Logger,
+		poolResource: params.PoolResource,
+		svcResource:  params.SvcResource,
+		poolClient:   params.Clientset.CiliumV2alpha1().CiliumLoadBalancerIPPools(),
+		svcClient:    params.Clientset.Slim().CoreV1(),
+		shutdowner:   params.Shutdowner,
+		pools:        make(map[string]*cilium_api_v2alpha1.CiliumLoadBalancerIPPool),
+		rangesStore:  newRangesStore(),
+		serviceStore: NewServiceStore(),
+		lbClasses:    lbClasses,
+		ipv4Enabled:  option.Config.IPv4Enabled(),
+		ipv6Enabled:  option.Config.IPv6Enabled(),
+	}
+
+	params.LC.Append(lbIPAM)
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	logger logrus.FieldLogger
+
+	lbClasses   []string
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*slim_core_v1.Service]
+
+	shutdowner hive.Shutdowner
+
+	pools        map[string]*cilium_api_v2alpha1.CiliumLoadBalancerIPPool
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// Only used during testing.
+	initDoneCallbacks []func()
+}
+
+// Start implements hive.HookInterface
+func (ipam *LBIPAM) Start(_ hive.HookContext) error {
+	ipam.logger.Info("Starting LB IPAM")
+
+	ipam.wg.Add(1)
+
+	// Start go routine, we can't block in OnStart
```

Comment:
Either you mean "in Start" here, or this comment is missing context. What is "OnStart" ? If its not visible here, it needs to be removed or further explained.